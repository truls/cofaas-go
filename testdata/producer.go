package main

import (
	"context"
	"flag"
	"fmt"
	storage "github.com/vhive-serverless/vSwarm/utils/storage/go"
	"math/rand"
	"net"
	"os"
	"strconv"

	sdk "github.com/ease-lab/vhive-xdt/sdk/golang"
	"github.com/ease-lab/vhive-xdt/utils"
	"google.golang.org/grpc/credentials/insecure"

	ctrdlog "github.com/containerd/containerd/log"
	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc/reflection"

	pb_client "tests/chained-functions-serving/proto"

	pb "github.com/vhive-serverless/vSwarm/examples/protobuf/helloworld"
	"google.golang.org/grpc"

	tracing "github.com/vhive-serverless/vSwarm/utils/tracing/go"
)

type producerServer struct {
	consumerAddr   string
	consumerPort   int
	payloadData    []byte
	XDTclient      *sdk.XDTclient
	transferType   string
	randomStr      string
	storageBackend storage.Storage
	pb.UnimplementedGreeterServer
}

type ubenchServer struct {
	consumerAddr   string
	consumerPort   int
	transferType   string
	payloadData    []byte
	XDTclient      *sdk.XDTclient
	randomStr      string
	storageBackend storage.Storage
	pb_client.UnimplementedProdConDriverServer
}

const (
	INLINE      = "INLINE"
	XDT         = "XDT"
	S3          = "S3"
	ELASTICACHE = "ELASTICACHE"
)

func uploadToStorage(ctx context.Context, payloadData []byte, randomStr string, storageBackend storage.Storage) string {
	span := tracing.Span{SpanName: "Storage put", TracerName: "Storage put - tracer"}
	span.StartSpan(ctx)
	defer span.EndSpan()

	key := fmt.Sprintf("payload_bytes_%s.txt", randomStr)
	key = storageBackend.Put(ctx, key, payloadData)
	log.Infof("[producer] Successfully uploaded %q", key)
	return key
}
func getGRPCclient(addr string) (pb_client.ProducerConsumerClient, *grpc.ClientConn) {
	// establish a connection
	var conn *grpc.ClientConn
	var err error
	if tracing.IsTracingEnabled() {
		conn, err = tracing.DialGRPCWithUnaryInterceptor(addr, grpc.WithBlock(), grpc.WithTransportCredentials(insecure.NewCredentials()))
	} else {
		conn, err = grpc.Dial(addr, grpc.WithBlock(), grpc.WithTransportCredentials(insecure.NewCredentials()))
	}
	if err != nil {
		log.Fatalf("[producer] fail to dial: %s", err)
	}
	return pb_client.NewProducerConsumerClient(conn), conn
}

func (ps *producerServer) SayHello(ctx context.Context, req *pb.HelloRequest) (_ *pb.HelloReply, err error) {
	addr := fmt.Sprintf("%v:%v", ps.consumerAddr, ps.consumerPort)
	if ps.transferType == INLINE || ps.transferType == S3 || ps.transferType == ELASTICACHE {
		client, conn := getGRPCclient(addr)
		defer conn.Close()
		payloadToSend := ps.payloadData
		if ps.transferType == S3 || ps.transferType == ELASTICACHE {
			payloadToSend = []byte(uploadToStorage(ctx, ps.payloadData, ps.randomStr, ps.storageBackend))
		}
		ack, err := client.ConsumeByte(ctx, &pb_client.ConsumeByteRequest{Value: payloadToSend})
		if err != nil {
			log.Fatalf("[producer] client error in string consumption: %s", err)
		}
		log.Printf("[producer] (single) Ack: %v\n", ack.Value)
	} else if ps.transferType == XDT {
		payloadToSend := utils.Payload{
			FunctionName: "HelloXDT",
			Data:         ps.payloadData,
		}
		if _, _, err := ps.XDTclient.Invoke(ctx, addr, payloadToSend); err != nil {
			log.Fatalf("SQP_to_dQP_data_transfer failed %v", err)
		}
	}
	return &pb.HelloReply{Message: "Success"}, err
}

func main() {
	flagAddress := flag.String("addr", "consumer.default.192.168.1.240.sslip.io", "Server IP address")
	flagClientPort := flag.Int("pc", 80, "Client Port")
	flagServerPort := flag.Int("ps", 80, "Server Port")
	url := flag.String("zipkin", "http://zipkin.istio-system.svc.cluster.local:9411/api/v2/spans", "zipkin url")
	dockerCompose := flag.Bool("dockerCompose", false, "Env docker Compose?")
	flag.Parse()

	log.SetFormatter(&log.TextFormatter{
		TimestampFormat: ctrdlog.RFC3339NanoFixed,
		FullTimestamp:   true,
	})
	log.SetOutput(os.Stdout)

	if tracing.IsTracingEnabled() {
		log.Println("producer has tracing enabled")
		shutdown, err := tracing.InitBasicTracer(*url, "producer")
		if err != nil {
			log.Warn(err)
		}
		defer shutdown()
	} else {
		log.Println("producer has tracing DISABLED")
	}

	var grpcServer *grpc.Server
	if tracing.IsTracingEnabled() {
		grpcServer = tracing.GetGRPCServerWithUnaryInterceptor()
	} else {
		grpcServer = grpc.NewServer()
	}

	//client setup
	log.Printf("[producer] Client using address: %v:%d\n", *flagAddress, *flagClientPort)

	ps := producerServer{consumerAddr: *flagAddress, consumerPort: *flagClientPort}
	us := ubenchServer{consumerAddr: *flagAddress, consumerPort: *flagClientPort}

	transferType, ok := os.LookupEnv("TRANSFER_TYPE")
	if !ok {
		log.Infof("TRANSFER_TYPE not found, using INLINE transfer")
		transferType = INLINE
	}
	log.Infof("[producer] transfering via %s", transferType)
	ps.transferType = transferType
	us.transferType = transferType

	transferSizeKB := 4095
	if value, ok := os.LookupEnv("TRANSFER_SIZE_KB"); ok {
		if intValue, err := strconv.Atoi(value); err == nil {
			transferSizeKB = intValue
		} else {
			log.Infof("invalid TRANSFER_SIZE_KB: %s, using default %d", value, transferSizeKB)
		}
	}

	// 4194304 bytes is the limit by gRPC
	payloadData := make([]byte, transferSizeKB*1024)
	if _, err := rand.Read(payloadData); err != nil {
		log.Fatal(err)
	}
	ps.randomStr = os.Getenv("HOSTNAME")
	us.randomStr = ps.randomStr

	log.Infof("sending %d bytes to consumer", len(payloadData))
	ps.payloadData = payloadData
	us.payloadData = payloadData
	if transferType == XDT {
		config := utils.ReadConfig()
		if !*dockerCompose {
			config.SQPServerHostname = utils.FetchSelfIP()
		}
		xdtClient, err := sdk.NewXDTclient(config)
		if err != nil {
			log.Fatalf("InitXDT failed %v", err)
		}

		ps.XDTclient = xdtClient

		us.XDTclient = xdtClient
	} else if transferType == S3 || transferType == ELASTICACHE {

		bucketName, ok := os.LookupEnv("BUCKET_NAME")
		if !ok {
			bucketName = "vhive-prodcon-bench"
		}
		storageBackend := storage.New(transferType, bucketName)
		us.storageBackend = storageBackend
		ps.storageBackend = storageBackend
	}
	pb.RegisterGreeterServer(grpcServer, &ps)
	pb_client.RegisterProdConDriverServer(grpcServer, &us)
	reflection.Register(grpcServer)

	//server setup
	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *flagServerPort))
	if err != nil {
		log.Fatalf("[producer] failed to listen: %v", err)
	}

	log.Println("[producer] Server Started")

	if err := grpcServer.Serve(lis); err != nil {
		log.Fatalf("[producer] failed to serve: %s", err)
	}

}
