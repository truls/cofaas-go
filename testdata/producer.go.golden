package impl

import (
	"context"
	"flag"
	"fmt"
	"math/rand"
	"net"
	"os"
	"strconv"

	ctrdlog "github.com/containerd/containerd/log"
	log "github.com/sirupsen/logrus"

	pb_client "cofaas/protos/prodcon"

	pb "cofaas/protos/helloworld"
)

type producerServer struct {
	consumerAddr	string
	consumerPort	int
	payloadData	[]byte
	transferType	string
	randomStr	string
	pb.UnimplementedGreeterServer
}

const (
	INLINE		= "INLINE"
	XDT		= "XDT"
	S3		= "S3"
	ELASTICACHE	= "ELASTICACHE"
)

var verbose = flag.Bool("v", false, "Be verbose")

func getGRPCclient(addr string) (pb_client.ProducerConsumerClient, *grpc.ClientConn) {

	var conn *grpc.ClientConn
	var err error
	conn, err = grpc.Dial(addr, grpc.WithBlock(), grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		log.Fatalf("[producer] fail to dial: %s", err)
	}
	return pb_client.NewProducerConsumerClient(conn), conn
}

func (ps *producerServer) SayHello(ctx context.Context, req *pb.HelloRequest) (_ *pb.HelloReply, err error) {
	addr := fmt.Sprintf("%v:%v", ps.consumerAddr, ps.consumerPort)
	client, conn := getGRPCclient(addr)
	defer conn.Close()
	payloadToSend := ps.payloadData
	for i := 1; i <= 1; i++ {
		ack, err := client.ConsumeByte(ctx, &pb_client.ConsumeByteRequest{Value: payloadToSend})
		if err != nil {
			log.Fatalf("[producer] client error in string consumption: %s", err)
		}
		if *verbose {
			log.Printf("[producer] (single) Ack: %v\n", ack.Value)
		}
	}
	return &pb.HelloReply{Message: "Success"}, err
}

func Main() {
	flagAddress := flag.String("addr", "consumer.default.192.168.1.240.sslip.io", "Server IP address")
	flagClientPort := flag.Int("pc", 80, "Client Port")
	flagServerPort := flag.Int("ps", 80, "Server Port")
	flag.Parse()

	log.SetFormatter(&log.TextFormatter{
		TimestampFormat:	ctrdlog.RFC3339NanoFixed,
		FullTimestamp:		true,
	})
	log.SetOutput(os.Stdout)

	log.Println("producer has tracing DISABLED")

	var grpcServer *grpc.Server
	grpcServer = grpc.NewServer()

	log.Printf("[producer] Client using address: %v:%d\n", *flagAddress, *flagClientPort)

	ps := producerServer{consumerAddr: *flagAddress, consumerPort: *flagClientPort}
	transferType, ok := os.LookupEnv("TRANSFER_TYPE")
	if !ok {
		log.Infof("TRANSFER_TYPE not found, using INLINE transfer")
		transferType = INLINE
	}
	log.Infof("[producer] transfering via %s", transferType)
	ps.transferType = transferType

	transferSizeKB := 1
	if value, ok := os.LookupEnv("TRANSFER_SIZE_KB"); ok {
		if intValue, err := strconv.Atoi(value); err == nil {
			transferSizeKB = intValue
		} else {
			log.Infof("invalid TRANSFER_SIZE_KB: %s, using default %d", value, transferSizeKB)
		}
	}

	payloadData := make([]byte, transferSizeKB*1024)
	if _, err := rand.Read(payloadData); err != nil {
		log.Fatal(err)
	}
	ps.randomStr = os.Getenv("HOSTNAME")

	log.Infof("sending %d bytes to consumer", len(payloadData))
	ps.payloadData = payloadData
	pb.RegisterGreeterServer(grpcServer, &ps)
	reflection.Register(grpcServer)

	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *flagServerPort))
	if err != nil {
		log.Fatalf("[producer] failed to listen: %v", err)
	}

	log.Println("[producer] Server Started")

	if err := grpcServer.Serve(lis); err != nil {
		log.Fatalf("[producer] failed to serve: %s", err)
	}

}

import "github.com/truls/cofaas-go/stubs/grpc"
import "github.com/truls/cofaas-go/stubs/net"
